rfcs/063-catalogue_pipeline_lambdas/README.md

```diff
@@ -0,0 +1,139 @@
+
+  Error handling could use [Partial batch responses](https://docs.aws.amazon.com/lambda/latest/dg/services-sqs-errorhandling.html#services-sqs-batchfailurereporting).
+  
+- Lambda is generally considered to be cost-effective for workloads with intermittent or unpredictable traffic patterns (1ms for 1024MB: $0.0000000167).
+
+- Processing speed could be slowed down by [AWS Lambda cold starts](https://aws.amazon.com/blogs/compute/operating-lambda-performance-optimization-part-1/) but this can be optimised by provisioning concurrency.
+  
+- The concepts-pipeline has demonstrated the value of using AWS Lambdas in a data pipeline.
+
+
+### Transition from ECS to Lambda
+
+In terms of transition from ECS to lambdas to prevent downtime and allow a smooth transition:
+
+- Update deployment to allow deployment of existing container images to both ECS & Lambda runtime environments.
+- Move services over one at a time, testing that a new Lambda service works by dual-running it alongside its ECS counterpart without publishing messages downstream.
+- When satisfied AWS Lambda behaviour for a service has reached parity we can remove the ECS service and infrastructure.
+
+## Potential issues
+
+AWS Lambda has a different execution model than ECS, with SQS messages causing Lambda invocations triggered by an AWS Lambda Trigger (managed invocations) and configuration options to dictate how many messages a Lambda invocation handles, how many Lambdas can run at once, and how long any single Lambda can run for. In ECS, tasks are launched by autoscaling rules, those tasks then poll SQS for messages processing and deleting them as they are received. In ECS when there is no more work to do tasks are terminated based on autoscaling rules.
+
+Potential issues may arise around this change in exectution model, specifically:
+
+### Concurrency
+
+With ECS concurrency can be limited by the number of running tasks which provides a _rough_ restriction on the number of messages processed at once. With AWS Lambda concurrency can be specified exactly with ["reserved concurrency"](https://docs.aws.amazon.com/lambda/latest/dg/configuration-concurrency.html), and in combination with [batch size](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#sqs-polling-behavior) can tightly control the number of messages a Lambda processes.
+
+As we are moving from one model to another, we will need to calibrate concurrency for each service by estimation and experiment. 
+
+Concurrency is of particular concern where we are making network requests to external services that may not be configured to meet demand, for example Elasticsearch or RDS databases.
+
+### Batching
+
+Batching describes the number of messages handed to a Lambda to process in one invocation. For the catalogue pipeline we will be invoking Lambdas from non-FIFO SQS messages, and this has [specific restrictions around batch size](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html). 
```

Comment:
There is an interesting potential conflict here.

FIFO queues allow Content Based Deduplication, which could be an improvement on the current setup (because of the explosion downstream of the relation embedder), but if they can't provide enough messages for efficient bulk updates in ES, then we won't be able to take advantage of that feature.

Reply:
I'll add this point to the doc.